--- init.old
+++ init
@@ -428,9 +428,9 @@ fi
 
 # locate boot media and mount it
 ebegin "Mounting boot media"
-nlplug-findfs $cryptopts -p /sbin/mdev ${KOPT_debug_init:+-d} \
-	${KOPT_usbdelay:+-t $(( $KOPT_usbdelay * 1000 ))} \
-	$repoopts -a /tmp/apkovls
+mkdir -p /media/mmcblk0p1
+mount -o ro /dev/mmcblk0p1 /media/mmcblk0p1
+ls /media/mmcblk0p1/*.apkovl.tar.gz > /tmp/apkovls
 eend $?
 
 # early console?
