--- system.dts.old	2015-01-12 23:48:19.815231000 +0100
+++ system.dts	2015-01-12 23:48:19.815231000 +0100
@@ -21,6 +21,31 @@
 	} ;
 	ps7_ddr_0: memory@00100000 {
 		device_type = "memory";
-		reg = <0x0 0x20000000>;
+		reg = <0x0 0x1E000000>;
 	} ;
 };
+
+&ps7_ethernet_0 {
+	phy-handle = <&phy0>;
+	mdio {
+		phy0: phy@1 {
+			compatible = "lantiq,phy11g";
+			device_type = "ethernet-phy";
+			reg = <1>;
+		};
+	};
+};
+
+&ps7_i2c_0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	eep@50 {
+		compatible = "24c64";
+		reg = <0x50>;
+	};
+	eep@51 {
+		compatible = "24c64";
+		reg = <0x51>;
+	};
+};
+
