--- fsbl_hooks.c.old
+++ fsbl_hooks.c
@@ -34,6 +34,7 @@
 
 /************************** Function Prototypes ******************************/
 
+u32 SetMacAddress();
 
 /******************************************************************************
 * This function is the hook which will be called  before the bitstream download.
@@ -112,6 +113,7 @@ u32 FsblHookBeforeHandoff(void)
 	 * Errors to be stored in the status variable and returned
 	 */
 	fsbl_printf(DEBUG_INFO,"In FsblHookBeforeHandoff function \r\n");
+	Status = SetMacAddress();
 
 	return (Status);
 }
