diff -rupN old/linux-xlnx-xilinx-v2014.3/arch/arm/configs/xilinx_zynq_defconfig linux-xlnx-xilinx-v2014.3/arch/arm/configs/xilinx_zynq_defconfig
--- old/linux-xlnx-xilinx-v2014.3/arch/arm/configs/xilinx_zynq_defconfig
+++ linux-xlnx-xilinx-v2014.3/arch/arm/configs/xilinx_zynq_defconfig
@@ -142,6 +142,7 @@ CONFIG_XILINX_AXI_EMAC=y
 CONFIG_XILINX_PS_EMAC=y
 CONFIG_MARVELL_PHY=y
 CONFIG_VITESSE_PHY=y
+CONFIG_LANTIQ_PHY=y
 CONFIG_MDIO_BITBANG=y
 CONFIG_INPUT_SPARSEKMAP=y
 CONFIG_INPUT_EVDEV=y
@@ -268,3 +269,14 @@ CONFIG_RCU_CPU_STALL_TIMEOUT=60
 CONFIG_FONTS=y
 CONFIG_FONT_8x8=y
 CONFIG_FONT_8x16=y
+CONFIG_WIRELESS=y
+CONFIG_CFG80211=y
+CONFIG_MAC80211=y
+CONFIG_RTLWIFI=y
+CONFIG_RTL8192CU=y
+CONFIG_RT2X00=y
+CONFIG_RT2800USB=y
+CONFIG_RT2800USB_RT3573=y
+CONFIG_RT2800USB_RT53XX=y
+CONFIG_RT2800USB_RT55XX=y
+CONFIG_RT2800USB_UNKNOWN=y
diff -rupN old/linux-xlnx-xilinx-v2014.3/drivers/misc/eeprom/at24.c linux-xlnx-xilinx-v2014.3/drivers/misc/eeprom/at24.c
--- old/linux-xlnx-xilinx-v2014.3/drivers/misc/eeprom/at24.c
+++ linux-xlnx-xilinx-v2014.3/drivers/misc/eeprom/at24.c
@@ -508,6 +508,9 @@ static int at24_probe(struct i2c_client
 		 * is recommended anyhow.
 		 */
 		chip.page_size = 1;
+		if(!strncmp(id->name,"24c64",5)){
+			chip.page_size = 32;
+		}
 
 		/* update chipdata if OF is present */
 		at24_get_ofdata(client, &chip);
diff -rupN old/linux-xlnx-xilinx-v2014.3/drivers/net/phy/Kconfig linux-xlnx-xilinx-v2014.3/drivers/net/phy/Kconfig
--- old/linux-xlnx-xilinx-v2014.3/drivers/net/phy/Kconfig
+++ linux-xlnx-xilinx-v2014.3/drivers/net/phy/Kconfig
@@ -112,6 +112,11 @@ config MICREL_PHY
 	---help---
 	  Supports the KSZ9021, VSC8201, KS8001 PHYs.
 
+config LANTIQ_PHY
+	tristate "Driver for Lantiq PHYs"
+	---help---
+	  Supports the 11G and 22E PHYs.
+
 config FIXED_PHY
 	bool "Driver for MDIO Bus/PHY emulation with fixed speed/link PHYs"
 	depends on PHYLIB=y
diff -rupN old/linux-xlnx-xilinx-v2014.3/drivers/net/phy/Makefile linux-xlnx-xilinx-v2014.3/drivers/net/phy/Makefile
--- old/linux-xlnx-xilinx-v2014.3/drivers/net/phy/Makefile
+++ linux-xlnx-xilinx-v2014.3/drivers/net/phy/Makefile
@@ -24,6 +24,7 @@ obj-$(CONFIG_NATIONAL_PHY)	+= national.o
 obj-$(CONFIG_DP83640_PHY)	+= dp83640.o
 obj-$(CONFIG_STE10XP)		+= ste10Xp.o
 obj-$(CONFIG_MICREL_PHY)	+= micrel.o
+obj-$(CONFIG_LANTIQ_PHY)	+= lantiq.o
 obj-$(CONFIG_MDIO_OCTEON)	+= mdio-octeon.o
 obj-$(CONFIG_MICREL_KS8995MA)	+= spi_ks8995.o
 obj-$(CONFIG_AT803X_PHY)	+= at803x.o
diff -rupN old/linux-xlnx-xilinx-v2014.3/drivers/uio/uio_pdrv_genirq.c linux-xlnx-xilinx-v2014.3/drivers/uio/uio_pdrv_genirq.c
--- old/linux-xlnx-xilinx-v2014.3/drivers/uio/uio_pdrv_genirq.c
+++ linux-xlnx-xilinx-v2014.3/drivers/uio/uio_pdrv_genirq.c
@@ -148,7 +148,7 @@ static int uio_pdrv_genirq_probe(struct
 	if (!uioinfo->irq) {
 		ret = platform_get_irq(pdev, 0);
 		uioinfo->irq = ret;
-		if (ret == -ENXIO && pdev->dev.of_node)
+		if (ret == -EINVAL && pdev->dev.of_node)
 			uioinfo->irq = UIO_IRQ_NONE;
 		else if (ret < 0) {
 			dev_err(&pdev->dev, "failed to get IRQ\n");
