diff -rupN old/linux-xlnx-xilinx-v2016.4/arch/arm/configs/xilinx_zynq_defconfig linux-xlnx-xilinx-v2016.4/arch/arm/configs/xilinx_zynq_defconfig
--- old/linux-xlnx-xilinx-v2016.4/arch/arm/configs/xilinx_zynq_defconfig
+++ linux-xlnx-xilinx-v2016.4/arch/arm/configs/xilinx_zynq_defconfig
@@ -114,6 +114,7 @@ CONFIG_XILINX_PS_EMAC=y
 CONFIG_MARVELL_PHY=y
 CONFIG_VITESSE_PHY=y
 CONFIG_REALTEK_PHY=y
+CONFIG_LANTIQ_PHY=y
 CONFIG_MDIO_BITBANG=y
 CONFIG_INPUT_SPARSEKMAP=y
 CONFIG_INPUT_EVDEV=y
@@ -239,3 +240,132 @@ CONFIG_RCU_CPU_STALL_TIMEOUT=60
 CONFIG_FONTS=y
 CONFIG_FONT_8x8=y
 CONFIG_FONT_8x16=y
+CONFIG_OVERLAY_FS=y
+CONFIG_SQUASHFS=y
+CONFIG_SQUASHFS_XZ=y
+CONFIG_SYN_COOKIES=y
+CONFIG_WIRELESS=m
+CONFIG_CFG80211=m
+CONFIG_MAC80211=m
+CONFIG_RTLWIFI=m
+CONFIG_RTL8188EU=m
+CONFIG_RTL8192CU=m
+CONFIG_RT2X00=m
+CONFIG_RT2800USB=m
+CONFIG_RT2800USB_RT3573=y
+CONFIG_RT2800USB_RT53XX=y
+CONFIG_RT2800USB_RT55XX=y
+CONFIG_RT2800USB_UNKNOWN=y
+CONFIG_WL_MEDIATEK=m
+CONFIG_MT7601U=m
+CONFIG_ATH_CARDS=m
+CONFIG_ATH9K_HTC=m
+CONFIG_B43=m
+CONFIG_BRCMFMAC=m
+CONFIG_BRCMFMAC_USB=y
+CONFIG_PPP=m
+CONFIG_PPP_ASYNC=m
+CONFIG_PPP_BSDCOMP=m
+CONFIG_PPP_DEFLATE=m
+CONFIG_PPP_FILTER=y
+CONFIG_PPP_MULTILINK=y
+CONFIG_PPP_SYNC_TTY=m
+CONFIG_USB_USBNET=m
+CONFIG_USB_NET_CDC_EEM=m
+CONFIG_USB_NET_CDC_MBIM=m
+CONFIG_USB_NET_HUAWEI_CDC_NCM=m
+CONFIG_USB_NET_QMI_WWAN=m
+CONFIG_USB_SERIAL_OPTION=m
+CONFIG_NETFILTER=y
+CONFIG_NETFILTER_XT_MATCH_CONNTRACK=m
+CONFIG_NETFILTER_XT_MATCH_STATE=m
+CONFIG_NF_CONNTRACK=m
+CONFIG_NF_CONNTRACK_IPV4=m
+CONFIG_NF_NAT_IPV4=m
+CONFIG_NF_NAT_MASQUERADE_IPV4=m
+CONFIG_IP_NF_IPTABLES=m
+CONFIG_IP_NF_NAT=m
+CONFIG_IP_NF_TARGET_MASQUERADE=m
+CONFIG_IP_NF_FILTER=m
+CONFIG_IP_NF_MANGLE=m
+CONFIG_TUN=m
+CONFIG_INPUT_TOUCHSCREEN=y
+CONFIG_OF_TOUCHSCREEN=m
+CONFIG_TOUCHSCREEN_USB_COMPOSITE=m
+CONFIG_TOUCHSCREEN_USB_EGALAX=m
+CONFIG_TOUCHSCREEN_USB_PANJIT=m
+CONFIG_TOUCHSCREEN_USB_3M=m
+CONFIG_TOUCHSCREEN_USB_ITM=m
+CONFIG_TOUCHSCREEN_USB_ETURBO=m
+CONFIG_TOUCHSCREEN_USB_GUNZE=m
+CONFIG_TOUCHSCREEN_USB_DMC_TSC10=m
+CONFIG_TOUCHSCREEN_USB_IRTOUCH=m
+CONFIG_TOUCHSCREEN_USB_IDEALTEK=m
+CONFIG_TOUCHSCREEN_USB_GENERAL_TOUCH=m
+CONFIG_TOUCHSCREEN_USB_GOTOP=m
+CONFIG_TOUCHSCREEN_USB_JASTEC=m
+CONFIG_TOUCHSCREEN_USB_ELO=m
+CONFIG_TOUCHSCREEN_USB_E2I=m
+CONFIG_TOUCHSCREEN_USB_ZYTRONIC=m
+CONFIG_TOUCHSCREEN_USB_ETT_TC45USB=m
+CONFIG_TOUCHSCREEN_USB_NEXIO=m
+CONFIG_TOUCHSCREEN_USB_EASYTOUCH=m
+CONFIG_FB_SYS_FOPS=m
+CONFIG_FB_DEFERRED_IO=m
+CONFIG_FB_MODE_HELPERS=m
+CONFIG_FB_UDL=m
+CONFIG_FUSE_FS=m
+CONFIG_SPI_SPIDEV=m
+CONFIG_XILINX_PS_EMAC_HWTSTAMP=y
+CONFIG_PPS_CLIENT_GPIO=m
+CONFIG_PPS_CLIENT_LDISC=m
+CONFIG_USB_EHCI_TT_NEWSCHED=y
+CONFIG_SND_USB_AUDIO=m
+CONFIG_USB_SERIAL=m
+CONFIG_USB_SERIAL_FTDI_SIO=m
+CONFIG_CIFS=m
+CONFIG_CIFS_STATS=y
+CONFIG_CIFS_STATS2=y
+CONFIG_CIFS_UPCALL=y
+CONFIG_CIFS_XATTR=y
+CONFIG_CIFS_POSIX=y
+CONFIG_CIFS_ACL=y
+CONFIG_CIFS_DFS_UPCALL=y
+CONFIG_CIFS_SMB2=y
+CONFIG_CIFS_FSCACHE=y
+CONFIG_OF_OVERLAY=y
+CONFIG_STAGING=y
+CONFIG_FB_TFT=m
+CONFIG_FB_TFT_AGM1264K_FL=m
+CONFIG_FB_TFT_BD663474=m
+CONFIG_FB_TFT_HX8340BN=m
+CONFIG_FB_TFT_HX8347D=m
+CONFIG_FB_TFT_HX8353D=m
+CONFIG_FB_TFT_HX8357D=m
+CONFIG_FB_TFT_ILI9163=m
+CONFIG_FB_TFT_ILI9320=m
+CONFIG_FB_TFT_ILI9325=m
+CONFIG_FB_TFT_ILI9340=m
+CONFIG_FB_TFT_ILI9341=m
+CONFIG_FB_TFT_ILI9481=m
+CONFIG_FB_TFT_ILI9486=m
+CONFIG_FB_TFT_PCD8544=m
+CONFIG_FB_TFT_RA8875=m
+CONFIG_FB_TFT_S6D02A1=m
+CONFIG_FB_TFT_S6D1121=m
+CONFIG_FB_TFT_SSD1289=m
+CONFIG_FB_TFT_SSD1306=m
+CONFIG_FB_TFT_SSD1331=m
+CONFIG_FB_TFT_SSD1351=m
+CONFIG_FB_TFT_ST7735R=m
+CONFIG_FB_TFT_ST7789V=m
+CONFIG_FB_TFT_TINYLCD=m
+CONFIG_FB_TFT_TLS8204=m
+CONFIG_FB_TFT_UC1611=m
+CONFIG_FB_TFT_UC1701=m
+CONFIG_FB_TFT_UPD161704=m
+CONFIG_FB_TFT_WATTEROTT=m
+CONFIG_FB_FLEX=m
+CONFIG_FB_TFT_FBTFT_DEVICE=m
+CONFIG_USB_ACM=m
+CONFIG_USB_TMC=m
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/iio/adc/xilinx-xadc-core.c linux-xlnx-xilinx-v2016.4/drivers/iio/adc/xilinx-xadc-core.c
--- old/linux-xlnx-xilinx-v2016.4/drivers/iio/adc/xilinx-xadc-core.c
+++ linux-xlnx-xilinx-v2016.4/drivers/iio/adc/xilinx-xadc-core.c
@@ -1090,7 +1090,7 @@ static int xadc_parse_dt(struct iio_dev
 	unsigned int num_channels;
 	const char *external_mux;
 	u32 ext_mux_chan;
-	int reg;
+	u32 reg;
 	int ret;
 
 	*conf = 0;
@@ -1142,13 +1142,13 @@ static int xadc_parse_dt(struct iio_dev
 				break;
 			}
 
-			channels[num_channels] = xadc_channels[reg + 9];
-			channels[num_channels].channel = num_channels - 1;
-
 			ret = of_property_read_u32(child, "reg", &reg);
 			if (ret || reg > 16)
 				continue;
 
+			channels[num_channels] = xadc_channels[reg + 9];
+			channels[num_channels].channel = num_channels - 1;
+
 			if (of_property_read_bool(child, "xlnx,bipolar"))
 				channels[num_channels].scan_type.sign = 's';
 
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/misc/eeprom/at24.c linux-xlnx-xilinx-v2016.4/drivers/misc/eeprom/at24.c
--- old/linux-xlnx-xilinx-v2016.4/drivers/misc/eeprom/at24.c
+++ linux-xlnx-xilinx-v2016.4/drivers/misc/eeprom/at24.c
@@ -508,6 +508,9 @@ static int at24_probe(struct i2c_client
 		 * is recommended anyhow.
 		 */
 		chip.page_size = 1;
+		if(!strncmp(id->name,"24c64",5)){
+			chip.page_size = 32;
+		}
 
 		/* update chipdata if OF is present */
 		at24_get_ofdata(client, &chip);
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/phy/Kconfig linux-xlnx-xilinx-v2016.4/drivers/net/phy/Kconfig
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/phy/Kconfig
+++ linux-xlnx-xilinx-v2016.4/drivers/net/phy/Kconfig
@@ -156,6 +156,11 @@ config MICROCHIP_PHY
 	help
 	  Supports the LAN88XX PHYs.
 
+config LANTIQ_PHY
+	tristate "Driver for Lantiq PHYs"
+	---help---
+	  Supports the 11G and 22E PHYs.
+
 config FIXED_PHY
 	tristate "Driver for MDIO Bus/PHY emulation with fixed speed/link PHYs"
 	depends on PHYLIB
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile linux-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
+++ linux-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
@@ -30,6 +30,7 @@ obj-$(CONFIG_DP83848_PHY)	+= dp83848.o
 obj-$(CONFIG_DP83867_PHY)	+= dp83867.o
 obj-$(CONFIG_STE10XP)		+= ste10Xp.o
 obj-$(CONFIG_MICREL_PHY)	+= micrel.o
+obj-$(CONFIG_LANTIQ_PHY)	+= lantiq.o
 obj-$(CONFIG_MDIO_OCTEON)	+= mdio-octeon.o
 obj-$(CONFIG_MDIO_THUNDER)	+= mdio-thunder.o
 obj-$(CONFIG_MDIO_CAVIUM)	+= mdio-cavium.o
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Kconfig linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Kconfig
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Kconfig
+++ linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Kconfig
@@ -12,7 +12,7 @@ config WLAN_VENDOR_REALTEK
 if WLAN_VENDOR_REALTEK
 
 source "drivers/net/wireless/realtek/rtl818x/Kconfig"
-source "drivers/net/wireless/realtek/rtlwifi/Kconfig"
-source "drivers/net/wireless/realtek/rtl8xxxu/Kconfig"
+source "drivers/net/wireless/realtek/rtl8188eu/Kconfig"
+source "drivers/net/wireless/realtek/rtl8192cu/Kconfig"
 
 endif # WLAN_VENDOR_REALTEK
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Makefile linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Makefile
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Makefile
+++ linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/Makefile
@@ -4,6 +4,6 @@
 
 obj-$(CONFIG_RTL8180)		+= rtl818x/
 obj-$(CONFIG_RTL8187)		+= rtl818x/
-obj-$(CONFIG_RTLWIFI)		+= rtlwifi/
-obj-$(CONFIG_RTL8XXXU)		+= rtl8xxxu/
+obj-$(CONFIG_RTL8188EU)		+= rtl8188eu/
+obj-$(CONFIG_RTL8192CU)		+= rtl8192cu/
 
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8188eu/Kconfig linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8188eu/Kconfig
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8188eu/Kconfig
+++ linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8188eu/Kconfig
@@ -0,0 +1,9 @@
+config RTL8188EU
+	tristate "Realtek 8188EU USB WiFi"
+	depends on MAC80211 && USB
+	select CFG80211_WEXT
+	select WIRELESS_EXT
+	select WEXT_PRIV
+	---help---
+	  This option adds the Realtek RTL8188EU USB device.
+
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8192cu/Kconfig linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8192cu/Kconfig
--- old/linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8192cu/Kconfig
+++ linux-xlnx-xilinx-v2016.4/drivers/net/wireless/realtek/rtl8192cu/Kconfig
@@ -0,0 +1,9 @@
+config RTL8192CU
+	tristate "Realtek 8192CU USB WiFi"
+	depends on MAC80211 && USB
+	select CFG80211_WEXT
+	select WIRELESS_EXT
+	select WEXT_PRIV
+	---help---
+	  This option adds the Realtek RTL8192CU USB device.
+
diff -rupN old/linux-xlnx-xilinx-v2016.4/drivers/pps/clients/pps-gpio.c linux-xlnx-xilinx-v2016.4/drivers/pps/clients/pps-gpio.c
--- old/linux-xlnx-xilinx-v2016.4/drivers/pps/clients/pps-gpio.c
+++ linux-xlnx-xilinx-v2016.4/drivers/pps/clients/pps-gpio.c
@@ -119,6 +119,8 @@ static int pps_gpio_probe(struct platfor
 
 		if (of_get_property(np, "assert-falling-edge", NULL))
 			data->assert_falling_edge = true;
+		if (of_get_property(np, "capture-clear", NULL))
+			data->capture_clear = true;
 	}
 
 	/* GPIO setup */
