diff -rupN old/u-boot-xlnx-xilinx-v2019.2/arch/arm/dts/Makefile u-boot-xlnx-xilinx-v2019.2/arch/arm/dts/Makefile
--- old/u-boot-xlnx-xilinx-v2019.2/arch/arm/dts/Makefile
+++ u-boot-xlnx-xilinx-v2019.2/arch/arm/dts/Makefile
@@ -164,7 +164,8 @@ dtb-$(CONFIG_ARCH_ZYNQ) += \
 	zynq-zed.dtb \
 	zynq-zturn.dtb \
 	zynq-zybo.dtb \
-	zynq-zybo-z7.dtb
+	zynq-zybo-z7.dtb \
+	zynq-red-pitaya.dtb
 dtb-$(CONFIG_ARCH_ZYNQMP) += \
 	avnet-ultra96-rev1.dtb			\
 	zynqmp-a2197-revA.dtb			\
diff -rupN old/u-boot-xlnx-xilinx-v2019.2/common/main.c u-boot-xlnx-xilinx-v2019.2/common/main.c
--- old/u-boot-xlnx-xilinx-v2019.2/common/main.c
+++ u-boot-xlnx-xilinx-v2019.2/common/main.c
@@ -58,6 +58,8 @@ void main_loop(void)
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
 
+	env_set("sdboot", "echo Importing environment from SD... && mmcinfo && load mmc 0 0x2000000 uEnv.txt && env import -t 0x2000000 ${filesize} && boot");
+
 	autoboot_command(s);
 
 	cli_loop();
diff -rupN old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Kconfig u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Kconfig
--- old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Kconfig
+++ u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Kconfig
@@ -106,6 +106,9 @@ config PHY_DAVICOM
 config PHY_ET1011C
 	bool "LSI TruePHY ET1011C support"
 
+config PHY_LANTIQ
+	bool "Lantiq Ethernet PHYs support"
+
 config PHY_LXT
 	bool "LXT971 Ethernet PHY support"
 
diff -rupN old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Makefile u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Makefile
--- old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Makefile
+++ u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/Makefile
@@ -16,6 +16,7 @@ obj-$(CONFIG_PHY_BROADCOM) += broadcom.o
 obj-$(CONFIG_PHY_CORTINA) += cortina.o
 obj-$(CONFIG_PHY_DAVICOM) += davicom.o
 obj-$(CONFIG_PHY_ET1011C) += et1011c.o
+obj-$(CONFIG_PHY_LANTIQ) += lantiq.o
 obj-$(CONFIG_PHY_LXT) += lxt.o
 obj-$(CONFIG_PHY_MARVELL) += marvell.o
 obj-$(CONFIG_PHY_MICREL_KSZ8XXX) += micrel_ksz8xxx.o
diff -rupN old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/phy.c u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/phy.c
--- old/u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/phy.c
+++ u-boot-xlnx-xilinx-v2019.2/drivers/net/phy/phy.c
@@ -503,6 +503,9 @@ int phy_init(void)
 #ifdef CONFIG_PHY_ET1011C
 	phy_et1011c_init();
 #endif
+#ifdef CONFIG_PHY_LANTIQ
+	phy_lantiq_init();
+#endif
 #ifdef CONFIG_PHY_LXT
 	phy_lxt_init();
 #endif
diff -rupN old/u-boot-xlnx-xilinx-v2019.2/include/phy.h u-boot-xlnx-xilinx-v2019.2/include/phy.h
--- old/u-boot-xlnx-xilinx-v2019.2/include/phy.h
+++ u-boot-xlnx-xilinx-v2019.2/include/phy.h
@@ -51,7 +51,7 @@
 				SUPPORTED_10000baseT_Full)
 
 #ifndef PHY_ANEG_TIMEOUT
-#define PHY_ANEG_TIMEOUT	4000
+#define PHY_ANEG_TIMEOUT	10000
 #endif
 
 
@@ -230,6 +230,7 @@ int phy_atheros_init(void);
 int phy_broadcom_init(void);
 int phy_cortina_init(void);
 int phy_davicom_init(void);
+int phy_lantiq_init(void);
 int phy_et1011c_init(void);
 int phy_lxt_init(void);
 int phy_marvell_init(void);
