diff -rupN old/u-boot-xlnx-xilinx-v2015.1/arch/arm/cpu/armv7/zynq/Kconfig u-boot-xlnx-xilinx-v2015.1/arch/arm/cpu/armv7/zynq/Kconfig
--- old/u-boot-xlnx-xilinx-v2015.1/arch/arm/cpu/armv7/zynq/Kconfig
+++ u-boot-xlnx-xilinx-v2015.1/arch/arm/cpu/armv7/zynq/Kconfig
@@ -24,6 +24,9 @@ config TARGET_ZYNQ_AFX
 config TARGET_ZYNQ_CSE
 	bool "Zynq CSE Board"
 
+config TARGET_ZYNQ_RED_PITAYA
+	bool "Red Pitaya Board"
+
 endchoice
 
 config SYS_BOARD
@@ -43,5 +46,6 @@ config SYS_CONFIG_NAME
 	default "zynq_zybo" if TARGET_ZYNQ_ZYBO
 	default "zynq_cse" if TARGET_ZYNQ_CSE
 	default "zynq_afx" if TARGET_ZYNQ_AFX
+	default "zynq_red_pitaya" if TARGET_ZYNQ_RED_PITAYA
 
 endif
diff -rupN old/u-boot-xlnx-xilinx-v2015.1/common/main.c u-boot-xlnx-xilinx-v2015.1/common/main.c
--- old/u-boot-xlnx-xilinx-v2015.1/common/main.c
+++ u-boot-xlnx-xilinx-v2015.1/common/main.c
@@ -76,6 +76,9 @@ void main_loop(void)
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
 
+	setenv("fdt_high", "0x1E000000");
+	setenv("sdboot", "echo Importing environment from SD... && mmcinfo && fatload mmc 0 0x2000000 uEnv.txt && env import -t 0x2000000 ${filesize} && boot");
+
 	autoboot_command(s);
 
 	cli_loop();
diff -rupN old/u-boot-xlnx-xilinx-v2015.1/include/phy.h u-boot-xlnx-xilinx-v2015.1/include/phy.h
--- old/u-boot-xlnx-xilinx-v2015.1/include/phy.h
+++ u-boot-xlnx-xilinx-v2015.1/include/phy.h
@@ -33,7 +33,7 @@
 				SUPPORTED_10000baseT_Full)
 
 #ifndef PHY_ANEG_TIMEOUT
-#define PHY_ANEG_TIMEOUT	4000
+#define PHY_ANEG_TIMEOUT	10000
 #endif
 
 
