diff -rupN old/u-boot-xlnx-xilinx-v2016.4/arch/arm/dts/Makefile u-boot-xlnx-xilinx-v2016.4/arch/arm/dts/Makefile
--- old/u-boot-xlnx-xilinx-v2016.4/arch/arm/dts/Makefile
+++ u-boot-xlnx-xilinx-v2016.4/arch/arm/dts/Makefile
@@ -97,7 +97,8 @@ dtb-$(CONFIG_ARCH_ZYNQ) += zynq-zc702.dt
 	zynq-zc770-xm010.dtb \
 	zynq-zc770-xm011.dtb \
 	zynq-zc770-xm012.dtb \
-	zynq-zc770-xm013.dtb
+	zynq-zc770-xm013.dtb \
+	zynq-red-pitaya.dtb
 dtb-$(CONFIG_ARCH_ZYNQMP) += \
 	zynqmp-ep108.dtb			\
 	zynqmp-zcu100.dtb			\
diff -rupN old/u-boot-xlnx-xilinx-v2016.4/common/main.c u-boot-xlnx-xilinx-v2016.4/common/main.c
--- old/u-boot-xlnx-xilinx-v2016.4/common/main.c
+++ u-boot-xlnx-xilinx-v2016.4/common/main.c
@@ -63,6 +63,9 @@ void main_loop(void)
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
 
+	setenv("fdt_high", "0x1E000000");
+	setenv("sdboot", "echo Importing environment from SD... && mmcinfo && fatload mmc 0 0x2000000 uEnv.txt && env import -t 0x2000000 ${filesize} && boot");
+
 	autoboot_command(s);
 
 	cli_loop();
diff -rupN old/u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
--- old/u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
+++ u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/Makefile
@@ -16,6 +16,7 @@ obj-$(CONFIG_PHY_ATHEROS) += atheros.o
 obj-$(CONFIG_PHY_BROADCOM) += broadcom.o
 obj-$(CONFIG_PHY_CORTINA) += cortina.o
 obj-$(CONFIG_PHY_DAVICOM) += davicom.o
+obj-$(CONFIG_PHY_LANTIQ) += lantiq.o
 obj-$(CONFIG_PHY_ET1011C) += et1011c.o
 obj-$(CONFIG_PHY_LXT) += lxt.o
 obj-$(CONFIG_PHY_MARVELL) += marvell.o
diff -rupN old/u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/phy.c u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/phy.c
--- old/u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/phy.c
+++ u-boot-xlnx-xilinx-v2016.4/drivers/net/phy/phy.c
@@ -482,6 +482,9 @@ int phy_init(void)
 #ifdef CONFIG_PHY_ET1011C
 	phy_et1011c_init();
 #endif
+#ifdef CONFIG_PHY_LANTIQ
+	phy_lantiq_init();
+#endif
 #ifdef CONFIG_PHY_LXT
 	phy_lxt_init();
 #endif
diff -rupN old/u-boot-xlnx-xilinx-v2016.4/include/linux/mii.h u-boot-xlnx-xilinx-v2016.4/include/linux/mii.h
--- old/u-boot-xlnx-xilinx-v2016.4/include/linux/mii.h
+++ u-boot-xlnx-xilinx-v2016.4/include/linux/mii.h
@@ -19,6 +19,8 @@
 #define MII_EXPANSION	    0x06	/* Expansion register	       */
 #define MII_CTRL1000	    0x09	/* 1000BASE-T control	       */
 #define MII_STAT1000	    0x0a	/* 1000BASE-T status	       */
+#define MII_MMD_CTRL	    0x0d	/* MMD Access Control Register */
+#define MII_MMD_DATA	    0x0e	/* MMD Access Data Register */
 #define MII_ESTATUS	    0x0f	/* Extended Status */
 #define MII_DCOUNTER	    0x12	/* Disconnect counter	       */
 #define MII_FCSCOUNTER	    0x13	/* False carrier counter       */
diff -rupN old/u-boot-xlnx-xilinx-v2016.4/include/phy.h u-boot-xlnx-xilinx-v2016.4/include/phy.h
--- old/u-boot-xlnx-xilinx-v2016.4/include/phy.h
+++ u-boot-xlnx-xilinx-v2016.4/include/phy.h
@@ -43,7 +43,7 @@
 				SUPPORTED_10000baseT_Full)
 
 #ifndef PHY_ANEG_TIMEOUT
-#define PHY_ANEG_TIMEOUT	4000
+#define PHY_ANEG_TIMEOUT	10000
 #endif
 
 
@@ -255,6 +255,7 @@ int phy_atheros_init(void);
 int phy_broadcom_init(void);
 int phy_cortina_init(void);
 int phy_davicom_init(void);
+int phy_lantiq_init(void);
 int phy_et1011c_init(void);
 int phy_lxt_init(void);
 int phy_marvell_init(void);
