--- old/wdsp-master/linux_port.c
+++ wdsp-master/linux_port.c
@@ -49,7 +49,7 @@ void InitializeCriticalSectionAndSpinCou
 	// DL1YCF: MacOS X does not have PTHREAD_MUTEX_RECURSIVE_NP
 	pthread_mutexattr_settype(&mAttr,PTHREAD_MUTEX_RECURSIVE);
 #else
-	pthread_mutexattr_settype(&mAttr,PTHREAD_MUTEX_RECURSIVE_NP);
+	pthread_mutexattr_settype(&mAttr,PTHREAD_MUTEX_RECURSIVE);
 #endif
 	pthread_mutex_init(mutex,&mAttr);
 	pthread_mutexattr_destroy(&mAttr);
