--- old/wdsp-master/linux_port.c
+++ wdsp-master/linux_port.c
@@ -45,7 +45,7 @@
 void InitializeCriticalSectionAndSpinCount(pthread_mutex_t *mutex,int count) {
 	pthread_mutexattr_t mAttr;
 	pthread_mutexattr_init(&mAttr);
-	pthread_mutexattr_settype(&mAttr,PTHREAD_MUTEX_RECURSIVE_NP);
+	pthread_mutexattr_settype(&mAttr,PTHREAD_MUTEX_RECURSIVE);
 	pthread_mutex_init(mutex,&mAttr);
 	pthread_mutexattr_destroy(&mAttr);
 	// ignore count
